File: src/shared/bump.ts
import { PLAYER_HALF, BOT_RADIUS } from "./constants.ts";

/** Post-bump immunity (fixed steps) — a RECONCILED tick gate: both sides run
 *  the identical countdown, so "can I be bumped?" is never a misprediction. */
export const BUMP_COOLDOWN_TICKS = 12;
export const BUMP_SPEED = 48;

export interface BumpableState { x: number; y: number; vx: number; vy: number; bumpTicks: number; }

/** Per-step cooldown countdown — run BEFORE the movement step on both sides. */
export function stepBumpGate(p: { bumpTicks: number }): void {
  if (p.bumpTicks > 0) p.bumpTicks--;
}

/**
 * Player-vs-bot bump test at an agreed bot position. Returns the knockback
 * velocity, or null. The caller applies it AND re-arms `bumpTicks` — on the
 * client that happens inside the reconciler step so the whole outcome
 * (velocity + immunity window) rides adopt+replay.
 */
export function collideBot(
  p: BumpableState,
  botX: number,
  botY: number,
): { vx: number; vy: number } | null {
  if (p.bumpTicks > 0) return null;
  const dx = p.x - botX, dy = p.y - botY;
  const r = PLAYER_HALF + BOT_RADIUS;
  const d2 = dx * dx + dy * dy;
  if (d2 >= r * r) return null;
  const d = Math.sqrt(d2) || 1e-6;
  return { vx: (dx / d) * BUMP_SPEED, vy: (dy / d) * BUMP_SPEED };
}

File: clients/flutter-app/lib/sim/bump.dart
import 'dart:math' as math;

import 'constants.dart';
import 'movement.dart';

/// Post-bump immunity (fixed steps) — a RECONCILED tick gate: both sides run
/// the identical countdown, so "can I be bumped?" is never a misprediction.
const int bumpCooldownTicks = 12;
const double bumpSpeed = 48.0;

/// Just the immunity counter, so [stepBumpGate] can run over anything carrying
/// one.
abstract class BumpGateState {
  int get bumpTicks;
  set bumpTicks(int value);
}

/// A body that can be knocked back.
abstract class BumpableState implements EntityState, BumpGateState {}

/// Per-step cooldown countdown — run BEFORE the movement step on both sides.
void stepBumpGate(BumpGateState p) {
  if (p.bumpTicks > 0) p.bumpTicks--;
}

/// Player-vs-bot bump test at an agreed bot position. Returns the knockback
/// velocity, or null.
///
/// The caller applies it AND re-arms `bumpTicks` — on the client that happens
/// inside the reconciler step so the whole outcome (velocity + immunity window)
/// rides adopt+replay.
({double vx, double vy})? collideBot(
  BumpableState p,
  double botX,
  double botY,
) {
  if (p.bumpTicks > 0) return null;
  final dx = p.x - botX, dy = p.y - botY;
  const r = playerHalf + botRadius;
  final d2 = dx * dx + dy * dy;
  if (d2 >= r * r) return null;
  var d = math.sqrt(d2);
  if (d == 0) d = 1e-6; // JS `|| 1e-6`
  return (vx: (dx / d) * bumpSpeed, vy: (dy / d) * bumpSpeed);
}
